### ISSUE TITLE
Difference in the loss computation between paper and code

### ISSUE THREAD

**Issue by EgShes** [yassouali/CCT#41]
I'm trying to re-implement your approach.

In the [paper](https://arxiv.org/pdf/2003.09005.pdf) in figure 3 you say _"The unsupervised loss is then computed between the outputs of the auxiliary decoders and that of the main decoder."_.

![image](https://user-images.githubusercontent.com/42002290/123088730-c5613b80-d42e-11eb-8464-83e44ea4307d.png)

But actually in the [model.py](https://github.com/yassouali/CCT/blob/master/models/model.py) (master, lines 114-125) you do the folloing:

```
# Get main prediction
x_ul = self.encoder(x_ul)
output_ul = self.main_decoder(x_ul)

# Get auxiliary predictions
outputs_ul = [aux_decoder(x_ul, output_ul.detach()) for aux_decoder in self.aux_decoders] 
targets = F.softmax(output_ul.detach(), dim=1)

# Compute unsupervised loss
loss_unsup = sum([
    self.unsuper_loss(inputs=u, targets=targets, conf_mask=self.confidence_masking, threshold=self.confidence_th, use_softmax=False)
    for u in outputs_ul
])
```

- compute main decoder output
- compute aux decoders output (can not understand why you are passing `output_ul.detach()` as it is ignored in aux_decoder's [forward](https://github.com/yassouali/CCT/blob/cf98ea7e6aefa7091e6c375a9025ba1e0f6e53ca/models/decoders.py#L72) )
- compute softmax for aux decoders outputs
- compute loss between each element in aux decoders output and softmaxed aux decoders output

Main decoder's output is not used in unsupervised loss computation as it is absent in `output_ul`. In [this](https://github.com/yassouali/CCT/blob/cf98ea7e6aefa7091e6c375a9025ba1e0f6e53ca/models/model.py#L119) line it is redefined

**Am I wrong** or your actual unsupervised loss computation is a little bit different than stated in the arxiv preprint?

**Comment by yassouali** [comment]
Hi @EgShes 

Thank you for your interest,

1- why you are passing output_ul.detach() ? 
The predictions are ignored in some cases, but for some types of aux decoders like ObjectMaskingDecoder the predictions are used for computing the perturbation.

2- Main decoder's output is not used in unsupervised loss:
In this unsupervised loss computation, we have `self.unsuper_loss(inputs=u, targets=targets)` where `u` is one of the auxiliary predictions, and `targets` are the main decoder's predictions since:
```python
x_ul = self.encoder(x_ul)
output_ul = self.main_decoder(x_ul)
targets = F.softmax(output_ul.detach(), dim=1)
```

Note that you can also do what you are describing, computing the losses between pairs of auxiliary decoder, this called pair wise loss and is explained in the sup. mat. of the paper.


Hope I answered your question.
